src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java

```diff
@@ -162,105 +169,201 @@ public int readClusters(long vcn, byte[] dst, int dstOffset, int nrClusters, int
+
+                // To process compressed buffers,
+                // the size of the compressed chunk that is stored in the chunk header MUST be used
+                // to determine the position of the last valid byte in the chunk.
+                // The size value MUST ignore flag bits that correspond to bytes outside the chunk.
+                if (cpos >= len + 3 || pos >= rightmostInUncompressed) {
+                    break;
+                }
+
                 if ((ctag & 1) != 0) {
                     int j, lmask, dshift;
                     for (j = pos - 1, lmask = 0xFFF, dshift = 12;
-                         j >= 0x10; j >>= 1) {
+                         j >= 0x10; j >>>= 1) {
                         dshift--;
-                        lmask >>= 1;
+                        lmask >>>= 1;
                     }
+
+                    // If the bit corresponding to a data element is set, the element is a two-byte compressed word.
                     final int tmp = compressed.getShort(cpos);
                     cpos += 2;
-                    final int boff = -1 - (tmp >> dshift);
-                    final int blen = Math.min(3 + (tmp & lmask), BLOCK_SIZE - pos);
 
-                    // Note that boff is negative.
-                    uncompressed.copyFrom(uncompressed, pos + boff, pos, blen);
-                    pos += blen;
+                    // While using the compressed buffers,
+                    // the stored displacement must be incremented by 1 and
+                    // the stored length must be incremented by 3,
+                    // to get the actual displacement and length.
+                    final int boff = (tmp >> dshift) + 1;
+                    int blen = (tmp & lmask) + 3;
+
+                    // Some of the bits in a flag byte might not be used.
+                    // To process compressed buffers,
+                    // the size of the compressed chunk that is stored in the chunk header MUST be used
+                    // to determine the position of the last valid byte in the chunk.
+                    // The size value MUST ignore flag bits that correspond to bytes outside the chunk.
+                    blen = Math.min(blen, rightmostInUncompressed - pos);
+
+                    // TODO,
+                    //  no idea why the offset may be even larger than the (uncompressed.array.offset + pos),
+                    //  -- it apparently makes it no place to start to read the data.
+                    //  we didn't find any documentation explaining this case..
+                    //  just return as an error for now.
+                    if (boff > uncompressed.offset + pos) {
+                        log.error("Failed to decompress data, the offset (to start to back to read) {} exceeds the sum of " +
+                                "the current position {} and the uncompressed.offset {}", boff, pos, uncompressed.offset);
+                        return pos;
```

Comment:
It seems like we should throw an exception here?  Logging and returning a valid RC will mean the upper-layers will continue to run as they think there is no error.